src/Dofs/sparsity_pattern.jl

```diff
@@ -102,38 +492,139 @@ function _coupling_to_local_dof_coupling(dh::DofHandler, coupling::AbstractMatri
+    heap = HeapAllocator.Heap()
+    sp′ = Dict{Int, HeapAllocator.HeapVector{Int}}()
+
+    for (row, colidxs) in zip(1:n_rows(sp), eachrow(sp)) # pairs(eachrow(sp))
+        row_coeffs = coefficients_for_dof(dofmapping, dofcoefficients, row)
+        if row_coeffs === nothing
+            # This row is _not_ constrained, check columns of this row...
+            !keep_constrained && haskey(dofmapping, row) && continue
+            for col in colidxs
+                col_coeffs = coefficients_for_dof(dofmapping, dofcoefficients, col)
+                if col_coeffs === nothing
+                    # ... this column is _not_ constrained, done.
+                    continue
+                else
+                    # ... this column _is_ constrained, distribute to columns.
+                    for (col′, _) in col_coeffs
+                        r = get(sp′, row) do
+                            HeapAllocator.resize(HeapAllocator.alloc_array(heap, Int, 8), 0)
+                        end
+                        r = insert_sorted(r, col′)
+                        sp′[row] = r
+                    end
+                end
+            end
+        else
+            # This row _is_ constrained, check columns of this row...
+            for col in colidxs
+                col_coeffs = coefficients_for_dof(dofmapping, dofcoefficients, col)
+                if col_coeffs === nothing
+                    # ... this column is _not_ constrained, distribute to rows.
+                    !keep_constrained && haskey(dofmapping, col) && continue
+                    for (row′, _) in row_coeffs
+                        r = get(sp′, row′) do
+                            HeapAllocator.resize(HeapAllocator.alloc_array(heap, Int, 8), 0)
+                        end
+                        r = insert_sorted(r, col)
+                        sp′[row′] = r
+                    end
+                else
+                    # ... this column _is_ constrained, double-distribute to columns/rows.
+                    for (row′, _) in row_coeffs
+                        !keep_constrained && haskey(dofmapping, row′) && continue
+                        for (col′, _) in col_coeffs
+                            !keep_constrained && haskey(dofmapping, col′) && continue
+                            r = get(sp′, row′) do
+                                HeapAllocator.resize(HeapAllocator.alloc_array(heap, Int, 8), 0)
+                            end
+                            r = insert_sorted(r, col′)
+                            sp′[row′] = r
+                        end
+                    end
+                end
+            end
+        end
+    end
+
+    # Insert new entries into the sparsity pattern
+    for (row, colidxs) in sp′
+        # TODO: Extract row here and just insert_sorted
+        for col in colidxs
+            add_entry!(sp, row, col)
+        end
+    end
+
+    # Release the memory in the temporary heap
+    HeapAllocator.free(heap)
```

Comment:
Since I just see it: Can't we automate this so this does not accidentally leak memory if e.g. one of the methods above throws an exception or errors? I have finalizers in mind (https://docs.julialang.org/en/v1/base/base/#Base.finalizer) but I am not sure if this works as intended.

Reply:
Yea I don't think this is necessary now after my simplifications.